### ISSUE TITLE
[Feature Request] Vim mode in debugger interface

### ISSUE THREAD

**Issue by erikw** [go-delve/delve#3384]
It would be great if fingers used to Vim-like navigation could command the delve debugger TUI e.g. like `<esc>kki` which would go up to lines in the history and enter insert mode again.

I'm used to this with other repl/debuggers like iPython, IRB, ghci etc.

Is delve using readline for the debugger TUI input? Then this would come for free when having `set editing-mode vi
` in `~/.inputrc` (did not work though so I assume not).

**Comment by aarzilli** [comment]
This was brought up before #2515

**Comment by itaranto** [comment]
This would be great, but I don't think delve uses `readline` so my guess is that it would have to be implemented from scratch.

Another alternative would be to write a custom client.

**Comment by itaranto** [comment]
I was taking a look at [liner](https://github.com/peterh/liner) which is the library used by `delve` to  implement the prompt.

Like already mentioned, there's an open [issue](https://github.com/peterh/liner/issues/120) for this. I took a look at the code and it seems implementing a Vi/Vim mode in `liner` would be huge pain in the ass.

So, I came across [this](https://github.com/lmorg/readline) library, which it seems to be a re-implementation of `readline` but in Go. It looks like it supports a Vim mode.

@aarzilli Have you considered using a different "liner" library?

**Comment by itaranto** [comment]
I just remember you guys at Delve use a mailing list, I'll try using `readline` for a bit and maybe I can create a proposal later.

**Comment by aarzilli** [comment]
We need ^C and ^Z to deliver signals in a specific way, these aren't tested automatically because we don't know how to write automated tests for it so it would have to be tested manually on every operating system we support (windows, macOS, linux and freebsd) and [bugs there can be subtle](https://github.com/go-delve/liner/commit/4726ab1d7f62f429ffd1bf4d2ad1e655dd12ced4).
That is to say: switching to a different readline library starts at a severe penalty and this issue isn't exactly bustling.

**Comment by aarzilli** [comment]
PS the library you are proposing has no tagged versions, a single importer and no activity for the past 3 years.

**Comment by itaranto** [comment]
> PS the library you are proposing has no tagged versions, a single importer and no activity for the past 3 years.

Yes, I know. But I looked at the code at it seems to be well structured compared to `liner`.  Also, the author seems to have "extracted" this code from another project of his ([murex](https://github.com/lmorg/murex)) which seems to be active.

To be fair, `liner` hasn't been updated in 3 years either.

**Comment by itaranto** [comment]
Oh, and BTW. The library *has* [tagged versions](https://github.com/lmorg/readline/tags) actually.

I think they don't show up in https://pkg.go.dev/ because of the Go version present in `go.mod`.

**Comment by itaranto** [comment]
Well I asked some [questions](https://github.com/lmorg/murex/discussions/891) to the Murex developer and it seems https://github.com/lmorg/readline it's not actually being maintained. Not outside Murex anyway.

I found another promising "readline" library which is actively maintained: https://github.com/reeflective/readline.

I did a PoC using that one and some (very) basic testing and it seems to work.
